### PR TITLE
Skip PP output communication for pure chunked prefill batches

### PR DESCRIPTION
**Pull request by xu-yfei**
## Motivation

In pipeline parallelism (PP), the last rank sends sampled `next_token_ids` to rank 0 after every forward pass. However, for chunked prefill requests that are **not** in their final chunk, `process_batch_result_prefill` only decrements a counter (`inflight_middle_chunks -= 1`) and completely discards the received tokens — they are never appended to `output_ids`.

When an entire microbatch consists solely of non-final chunked prefill requests, the output send/recv is pure overhead: the transmitted tensor is garbage data that no consumer will ever use. For long inputs that require many prefill chunks (e.g., 256K tokens with 8K chunk size = 32 chunks), this adds up to 31 unnecessary send/recv operations per request.

**Why this matters on H20-3E (78 SM per GPU)**: PP send and recv each occupy 1 SM, leaving only 77 SMs for compute. Kernels that require all 78 SMs to launch in a single wave are forced into two waves, roughly doubling their latency. For example, `flash_fwd_splitkv_mla_fp8_sparse_kernel` launches with grid 2×39 = 78, and its latency increases from 4.9ms to 9.7ms when only 77 SMs are available. The same effect is observed with `deep_gemm::sm90_fp8_paged_mqa_logits`, `nvjet_sm90_tst_128x256_64x4_2x1_v_bz_coopA_TNN`, and `deepgemm` — all launch with grid=78 and are similarly impacted.

image

image

image

image

image

## Modifications

- **`environ.py`**: Add `SGLANG_PP_SKIP_PURE_CHUNKED_OUTPUT_COMM = EnvBool(False)` — off by default.
- **`schedule_batch.py`**: Add `ScheduleBatch.contains_last_prefill_chunk: bool = True` (frozen at scheduling time, defaults `True` for safety) and `Req.inflight_middle_chunks` (replaces `is_chunked`).
- **`schedule_policy.py`**: `PrefillAdder` computes `contains_last_prefill_chunk` — set `True` when any request is a complete prefill or final chunk, left `False` when all are non-final chunks.
- **`scheduler.py`**: Copy `adder.contains_last_prefill_chunk` to `new_batch`; increment `inflight_middle_chunks` when dispatching a chunked request.
- **`scheduler_pp_mixin.py`**: Add `_pp_can_skip_output_comm(batch)` — returns `True` when env var is on, batch is extend (not prebuilt), `not contains_last_prefill_chunk`, and `not return_logprob`. Last-rank send path skips `send_dict_to_next_stage`; recv path calls `_pp_make_skip_output_result()` which returns placeholder zeros with `skipped_output_comm=True`. Both sender and receiver independently check the same batch object via their respective microbatch slot, so their conclusions are always consistent — no deadlock.
- **`utils.py`**: Add `skipped_output_comm: bool = False` to `GenerationBatchResult`.
- **`batch_result_processor.py`**: Add `_validate_pp_skip_output_comm()` — asserts `inflight_middle_chunks > 0` for all reqs when skip happened (placeholder zeros must not be consumed); warns when skip didn't happen but no req consumed output (missed opportunity). Gated by the same env var.

## Accuracy Tests

DeepSeek-V4-Pro-FP8, PP2, TP8, H20-3E, `SGLANG_PP_SKIP_PURE_CHUNKED_OUTPUT_COMM=1`, default chunked-prefill-size=8192.

| Benchmark | Score |
|-----------|-------|
| GSM8K 8-shot | 96.41% |
| MMLU | 90.78% |
| GPQA | 89.90% |

## Speed Tests and Profiling
DeepSeek-V4-Pro-FP8, PP2, TP8, H20-3E, non-DeepEP, default chunked-prefill-size=8192. TTFT: num-prompts=10, max-concurrency=1, random-output-len=256. TPOT: random-input-len=65536, random-output-len=1024. Values are minimum of 3 rounds.

### TTFT (ms)

| Input Len | Baseline | skip=1 | Delta |
|-----------|----------|--------|-------|
| 16K | 1268 | 1259 | -0.7% |
| 32K | 2231 | 2122 | -4.9% |
| 64K | 4086 | 3958 | -3.1% |
| 128K | 8285 | 7775 | -6.2% |
| 256K | 17444 | 16263 | -6.8% |

### TPOT (ms)

| Batch | Baseline | skip=1 | Delta |
|-------|----------|--------|-------|
| 1 | 19.38 | 19.04 | -1.8% |
| 4 | 30.06 | 30.13 | +0.2% |
| 8 | 38.01 | 37.31 | -1.8% |
| 16 | 58.86 | 56.89 | -3.3% |

TTFT improvement increases with input length (up to -6.8% at 256K) as longer inputs produce more non-final chunks. TPOT is neutral — skip only affects prefill, not decode.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #26443671822](https://github.com/sgl-project/sglang/actions/runs/26443671822)
Latest PR Test (Extra): :x: [Run #26443671690](https://github.com/sgl-project/sglang/actions/runs/26443671690)

### CODE REVIEW

**Comment by xu-yfei** [comment]
@ShangmingCai @whybeyoung Could you help review this?

**Review by ShangmingCai — COMMENTED** [review]
Nice PR, but except for assigning and passing contains_last_prefill_chunk through many modules, could we do in the req level, not at the batch level?

File: python/sglang/srt/managers/scheduler_pp_mixin.py [hunk 1]
```diff
@@ -41,6 +41,19 @@
...
 
+def _pp_can_skip_output_comm(batch: ScheduleBatch) -> bool:
+    """Check if output send/recv can be skipped for this batch."""
+    return (
+        envs.SGLANG_PP_SKIP_PURE_CHUNKED_OUTPUT_COMM.get()
+        and batch is not None
+        and batch.forward_mode.is_extend()
+        and not batch.forward_mode.is_prebuilt()
+        and len(batch.reqs) > 0
+        and not batch.contains_last_prefill_chunk
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/scheduler_pp_mixin.py:55]: Something like `all(req.inflight_middle_chunks > 0 for req in batch.reqs)`

**Review by ShangmingCai — APPROVED** [review]
LGTM

**Review by ShangmingCai — APPROVED** [review]
LGTM, now the logic is more concise, which also covers all cases.

**Comment by ShangmingCai** [comment]
/rerun-test test/registered/disaggregation/test_disaggregation_pp.py test/registered/pp/test_pp_single_node.py

**Comment by ShangmingCai** [comment]
/rerun-test test/registered/distributed/test_disaggregation_pp.py test/registered/distributed/test_pp_single_node.py

File: python/sglang/srt/managers/scheduler_components/batch_result_processor.py [hunk 2]
```diff
@@ -208,6 +208,8 @@ def process_batch_result_prefill(
             next_token_ids = next_token_ids.tolist()
             self._move_logprobs_to_cpu(batch=batch, logits_output=logits_output)
 
+            self._validate_pp_skip_output_comm(batch, result)
+
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/scheduler_components/batch_result_processor.py:212]: Should we check if pp size > 1 and disaggregation model == prefill, then we do this here? 

Since some CIs could not pass in the current state, and I could not update branch to this PR, the only thing I am concerning is this line, so if it is gated, we shall be able to only check the pp-related tests.

Comment (xu-yfei) [reply python/sglang/srt/managers/scheduler_components/batch_result_processor.py:212]: I have added environment variable control. When `SGLANG_PP_SKIP_PURE_CHUNKED_OUTPUT_COMM` is not enabled, it returns directly:

``` 
   @staticmethod
    def _validate_pp_skip_output_comm(
        batch: ScheduleBatch,
        result: Union[GenerationBatchResult, EmbeddingBatchResult],
    ):
        if not envs.SGLANG_PP_SKIP_PURE_CHUNKED_OUTPUT_COMM.get():
            return
```

**Comment by ShangmingCai** [comment]
/rerun-test test/registered/disaggregation/test_disaggregation_pp.py test/registered/pp/test_pp_single_node.py

**Comment by ShangmingCai** [comment]
/rerun-test test/registered/disaggregation/test_disaggregation_pp.py test/registered/pp/test_pp_single_node.py
